Add a CLI reference generator

ptgott opened the pull request:
See #3568

Add a function to update the kingpin usage template for a CLI application and print a docs page. As a starting point, includes a `docs` subcommand in `tsh` to call the function for that tool. By generating CLI documentation, we can ensure that commmand, argument, and flag usage information is consistent with the terminal help text for a given CLI. We can also ensure that all changes to a CLI application are reflected in the documentation.

As an alternative, we could edit `UpdateAppUsageTemplate` in `lib/utils` to update the usage template to print an MDX page, e.g., using build tags to print a docs template instead of the default usage template. This approach would update all CLI tools at once and use a consistent approach with minimal changes to each CLI tool. However, using a separate function to take a `*kingpin.Application` and print a docs page is more straightforward and explicit, even though it does require editing each CLI tool to call the function.

One shortcoming of generating the CLI reference docs versus our current manual approach is that it is non-trivial (and potentially impossible) to use `kingpin`'s usage API to obtain the value types of the subcommands, flags, and arguments registered against a `*kingpin.Application`. We can use flag and argument descriptions and default values to achieve the same purpose.

Another shortcoming is that `kingpin`'s usage API does not print usage information for arguments, environment variable, etc. that a CLI looks up directly from the OS, such as the `TELEPORT_CLUSTER` environment variable for `tsh`.

ptgott committed:
Add a CLI reference generator

See #3568

Add a function to update the kingpin usage template for a CLI
application and print a docs page. As a starting point, includes a
`docs` subcommand in `tsh` to call the function for that tool. By
generating CLI documentation, we can ensure that commmand, argument, and
flag usage information is consistent with the terminal help text for a
given CLI. We can also ensure that all changes to a CLI application are
reflected in the documentation.

As an alternative, we could edit `UpdateAppUsageTemplate` in `lib/utils`
to update the usage template to print an MDX page, e.g., using build
tags to print a docs template instead of the default usage template.
This approach would update all CLI tools at once and use a consistent
approach with minimal changes to each CLI tool. However, using a
separate function to take a `*kingpin.Application` and print a docs page
is more straightforward and explicit, even though it does require
editing each CLI tool to call the function.

One shortcoming of generating the CLI reference docs versus our current
manual approach is that it is non-trivial (and potentially impossible)
to use `kingpin`'s usage API to obtain the value types of the
subcommands, flags, and arguments registered against a
`*kingpin.Application`. We can use flag and argument descriptions and
default values to achieve the same purpose.

Another shortcoming is that `kingpin`'s usage API does not print usage
information for arguments, environment variable, etc. that a CLI looks
up directly from the OS, such as the `TELEPORT_CLUSTER` environment
variable for `tsh`.

ptgott commented:
To reduce the size of the diff, this PR doesn't add a Make target or the results of generating the reference page for `tsh`.

You can test it by building `tsh`, then running `tsh docs` and redirecting the output to, say, `docs-website/content/17.x/docs/pages/reference/cli/tsh-generated.mdx`, then running `yarn start` in your `docs-website` clone to start the dev server. Then navigate to `localhost:3000/docs/reference/cli/tsh-generated`.

atburke commented on lib/utils/docs.go:
```diff
@@ -0,0 +1,299 @@
+	return ret
+}
+
+// formatHelp prints help text to include in a Markdown table cell. It escapes
+// curly braces to avoid breaking the MDX parser.
+func formatHelp(help string) string {
+	return strings.NewReplacer("{", `\{`, "}", `\}`).Replace(help)
+}
```
I think we'll need to escape pipes too.
```suggestion
// formatHelp prints help text to include in a Markdown table cell. It escapes
// curly braces to avoid breaking the MDX parser, and it escapes pipes to
// avoid breaking the cell.
func formatHelp(help string) string {
	return strings.NewReplacer("{", `\{`, "}", `\}`, "|", `\|`).Replace(help)
}
```

ptgott replied:
Thanks!

zmb3 commented:
This may seem silly, but it feels wrong for us to ship this code (which is only needed to generate documentation) in builds of teleport, tsh, etc. 

I fear by putting it in a generic package like lib/utils that we will end up doing exactly that. 

Can we maybe hide this code behind a build tag so it doesn't make its way into production releases?

ptgott committed:
Respond to zmb3 and atburke feedback

**Separate docs generation code from production code.**

Add a separate implementation of `UpdateAppUsageTemplate` that requires
the `docs` build tag. Since all Go-based Teleport CLI tools call
`UpdateAppUsageTemplate`, the `docs` build of the function ensure that
the `help` command for Teleport CLI tools prints a docs page.

This change also adds a `cli-docs-tsh` make target to generate a docs
page for `tsh`.

**Escape pipes in `formatHelp`** to avoid breaking table cells.

ptgott commented on tool/tsh/common/tsh.go:
```diff
@@ -1737,6 +1739,8 @@ func Run(ctx context.Context, args []string, opts ...CliOption) error {
 		err = gitCmd.clone.run(&cf)
 	case pivCmd.agent.FullCommand():
 		err = pivCmd.agent.run(&cf)
+	case printDocs.FullCommand():
+		utils.PrintCLIDocs(os.Stdout, app)
```
Based on [Zac's feedback](https://github.com/gravitational/teleport/pull/54394#issuecomment-2843938991), I am going to take an approach that doesn't involve editing `tsh.go`, and instead use build tags to modify `UpdateAppUsageTemplate` to generate a docs page.

ptgott replied:
Ah, I see what you mean. Right now, the app-wide help text goes [under the usage example](https://github.com/gravitational/teleport/blob/a7c3c965212689081f15438b9fbc5f9d76c20d60/lib/utils/docs-usage.md.tmpl#L54), and I can move it up so it's more visible. One issue is that our `InitCLIParser` calls don't use complete sentences to describe our CLI tools ([e.g., for tsh](https://github.com/gravitational/teleport/blob/a7c3c965212689081f15438b9fbc5f9d76c20d60/tool/tsh/common/tsh.go#L796)). Would it make sense to edit these top-level help descriptions to be complete sentences, or do we need these to be short?

atburke replied:
Hmm. I think the short description is good for tsh. In the web docs we have more space for the extended description, and my hunch is that the longer description will be more useful to someone browsing the docs for the first time than someone who has already downloaded Teleport. If we want to keep both descriptions the same, I think the way you have it currently is a good balance. Up to you which you want to do.

ptgott replied:
Sounds good! I suppose we can leave it this way for now. If we want to add a hardcoded intro to each page, we could do something similar to the Terraform doc generator and add a directory of MD files that, if named after a particular CLI tool, we'd include in the page.

atburke commented on Makefile:
Can we make sure that tests under the `docs` build tag are exercised in the makefile somewhere? I worry it'll be easy to forget to run them if we don't.

ptgott committed:
Respond to atburke feedback

- Make the main help description more visible. Assume that the main help
  description is not necessarily a complete sentence, adding the text to
  the end of the introductory sentence using a colon.

- Add a make rule for running CLI docs generator tests. Use the `-run`
  flag since the tests are in a package we already run tests for in
  another make rule.

ptgott commented:
@zmb3 Just checking if there is anything left to do here. Thanks!

zmb3 commented on Makefile:
```diff
@@ -931,6 +931,14 @@ test-go-unit-tbot:
 		| tee $(TEST_LOG_DIR)/unit.json \
 		| gotestsum --raw-command -- cat
 
+# Runs CLI reference doc generator unit tests
+.PHONY: test-go-unit-clidocs
+test-go-unit-clidocs: FLAGS ?= -shuffle on -tags docs -run 'TestUpdateAppUsageTemplate'
+test-go-unit-clidocs:
+	go test -json $(FLAGS) ./lib/utils/... \
+		| tee $(TEST_LOG_DIR)/unit.json \
+		| gotestsum --raw-command -- cat
+
```
Do we need a separate target for this, or can we just pass the `docs` tag to the existing unit test target?

zmb3 commented on lib/utils/usage_docs.go:
```diff
@@ -0,0 +1,299 @@
+func formatThreeColMarkdownTable(rows [][3]string) string {
+	buf := bytes.NewBuffer(nil)
+	for _, r := range rows {
+		buf.WriteString("\n|" + r[0] + "|" + r[1] + "|" + r[2] + "|")
```
This could also be expressed as `fmt.Fprintf(buf, "\n|%v|%v|%v|", r[0], r[1], r[2])` if you like that better. No strong preference on my side.

zmb3 commented on lib/utils/usage_docs.go:
```diff
@@ -0,0 +1,299 @@
+// formatThreeColMarkdownTable formats the provided row data into a three-column
+// Markdown table, minus the header.
+func formatThreeColMarkdownTable(rows [][3]string) string {
+	buf := bytes.NewBuffer(nil)
```
```suggestion
	var buf bytes.Buffer
```

The zero-value for a `bytes.Buffer` works just fine, and then there are no questions about the mysterious `nil` value there.

zmb3 commented on lib/utils/usage_docs.go:
```diff
@@ -0,0 +1,299 @@
+			formatHelp(flag.Help),
+		})
+	}
+	return rows
```
Why is it called _flags to columns_ if we're returning a set of rows?

zmb3 commented on lib/utils/usage_docs.go:
```diff
@@ -0,0 +1,299 @@
+// anyVisibleFlags returns whether any flags in f are visible, i.e., should be
+// included in a table of flags.
+func anyVisibleFlags(f []*kingpin.FlagModel) bool {
+	for _, l := range f {
```
I'd use `slices.ContainsFunc` here.

zmb3 commented on lib/utils/usage_docs.go:
```diff
@@ -0,0 +1,299 @@
+// anyEnvVarsForCmd indicates whether at least one of the arguments and flags
+// provided exposes an environment variable for configuration.
+func anyEnvVarsForCmd(args []*kingpin.ArgModel, flags []*kingpin.FlagModel) bool {
+	for _, a := range args {
```
Similarly, `slices.ContainsFunc` would work nicely here.

zmb3 commented on lib/utils/usage_docs.go:
```diff
@@ -0,0 +1,299 @@
+		// Some commands declare empty argument names and help texts as
+		// a hack to allow arbitrary values. Indicate this in the table
+		// as a special case.
+		var argName string
+		if arg.Name != "" {
+			argName = arg.Name
+		} else {
+			argName = "args"
+		}
```
```suggestion
		argName := cmp.Or(arg.Name, "args")
```

ptgott replied:
Nice

zmb3 commented on lib/utils/usage_docs.go:
```diff
@@ -0,0 +1,299 @@
+			argName = "args"
+		}
+
+		var help string
+		if arg.Help != "" {
+			help = formatHelp(arg.Help)
+		} else {
+			help = "Arbitrary arguments"
+		}
```
```suggestion
		help := "Arbitrary arguments"
		if arg.Help != "" {
			help = formatHelp(arg.Help)
		}
```

zmb3 commented on lib/utils/usage_docs.go:
```diff
@@ -0,0 +1,299 @@
+	col := cmdModelCollection(cmds)
+	sort.Stable(&col)
+	return []*kingpin.CmdModel(col)
+}
```
I think you can probably replace this entire chunk with a single call to `slices.SortFunc`.

ptgott committed:
Respond to zmb3 feedback

- Use Fprintf instead of Buffer.WriteString.
- Use the zero value of bytes.Buffer instead of NewBuffer.
- Rename `.*ToColumns` functions to `.*ToRows` for clarity.
- Use slices.ContainsFunc in anyVisibleFlags.
- Use slices.ContainsFunc in anyEnvVarsForCmd.
- Clean up argsToRows.
- Minor cleanup in TestUpdateAppUsageTemplate.
- Clean up sortCommandsByName (no need to implement sort.Interface).
- Read the docs usage template from disk in UpdateAppUsageTemplate. For
  this to work, add the function updateAppUsageTemplate, which takes an
  io.Reader, and use separate file paths to pass the io.Reader to the
  function in the tests and in UpdateAppUsageTemplate.
- Inline formatFlagForTable and remove unnecessary branching.
- Add an ADDTAGS environment variable to the `test-go-unit` make
  target so it's possible to pass arbitrary build tags, e.g., "docs".
